ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,20 +187,407 @@ let order_f (_, vm_rec) =
+let update_anti_affinity_grp_ranked_hosts group host grp_ranked_hosts =
+  let ranked_hosts =
+    try VMGroupMap.find group grp_ranked_hosts
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, group not in it" ;
+      raise Not_found
+  in
+  let vm_cnt_on_host =
+    try
+      IntMap.find_first
+        (fun vm_cnt -> List.mem host (IntMap.find vm_cnt ranked_hosts))
+        ranked_hosts
+      |> fst
+    with Not_found ->
+      error "Failed to update grp_ranked_hosts, host not in it" ;
+      raise Not_found
+  in
+  VMGroupMap.update group
+    (fun _ ->
+      Some
+        (ranked_hosts
+        |> IntMap.update vm_cnt_on_host (Option.map (List.filter (( <> ) host)))
+        |> IntMap.update (vm_cnt_on_host + 1) (fun hs ->
+               Some (host :: Option.value ~default:[] hs)
+           )
+        )
+    )
+    grp_ranked_hosts
+
+(*****************************************************************************************************)
+(* Planning code follows                                                                             *)
+
+(** Try to get a spread evenly plan for anti-affinity VMs (for each anti-affinity group, the number of 
+    running VMs from that group are spread evenly in all the rest hosts in the pool):
+    1. For all the anti-affinity VMs which sorted in an increasing order in terms of the VM's memory
+       size, do host selection as below step 2.
+    2. For each anti-affinity VM, select a host which can run it, and which has the least VMs in the
+       same anti-affinity group running on it, for the hosts with the same number of running VMs in
+       that group, pick the one with the least free memory. *)
+let rec vm_anti_affinity_spread_evenly_plan ~__context hosts_increasing
+    vms_increasing vm_can_boot_on_host grp_ranked_hosts mapping =
+  debug "vm_anti_affinity_spread_evenly_plan" ;
+  log_vms_increasing ~__context vms_increasing ;
+  log_hosts_increasing ~__context hosts_increasing ;
+  log_grp_ranked_hosts ~__context grp_ranked_hosts ;
+  match vms_increasing with
+  | [] ->
+      mapping
+  | (vm, vm_size) :: remaining_vms -> (
+      let host_selector hosts =
+        hosts_increasing
+        |> List.filter (fun (h, size) -> List.mem h hosts && size >= vm_size)
+        |> List.map fst
+        |> fst_host_can_run vm_can_boot_on_host vm
+      in
+      match Xapi_vm_helpers.vm_has_anti_affinity ~__context ~vm with
+      | Some (`AntiAffinity group) -> (
+          debug
+            "Spread evenly plan: try to plan for the anti-affinity VM (%s, %s)"
+            (Ref.string_of vm)
+            (Db.VM.get_name_label ~__context ~self:vm) ;
+          try
+            let h =
+              Xapi_vm_helpers.select_host_from_ranked_groups vm host_selector
+                (VMGroupMap.find group grp_ranked_hosts
+                |> IntMap.to_seq
+                |> List.of_seq
+                |> List.fast_sort (fun (a, _) (b, _) -> compare a b)
```

Comment:
It's already ` in ascending order of keys`.

Reply:
I think you are referring to the [doc](https://ocaml.org/manual/5.1/api/Map.Make.html) of `Maps.to_seq`:
```
val to_seq : 'a t -> (key * 'a) Seq.t
Iterate on the whole map, in ascending order of keys
```
But I don't see the sequence is preserved by `List.of_seq`:
```
val of_seq : 'a Seq.t -> 'a list
Create a list from a sequence.
```
And from the name of `Maps.to_seq`, I am afraid the ascending order might not always be guaranteed, while it is required for us here, so adding a sort here might be better.

Reply:
The increasing order is only needed for the ranks, thanks, changed to use `Map.bindings`.